File: framework/languages/java/samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/configuration/RedisSpotRemoteAddressResolver.java
package systems.zlink.samples.tictactoe.server.configuration;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import systems.zlink.contracts.core.RoutingId;
import systems.zlink.framework.spots.ZLinkSpotKind;
import systems.zlink.framework.spots.ZLinkSpotRemoteAddress;
import systems.zlink.framework.spots.ZLinkSpotRemoteAddressResolver;

public final class RedisSpotRemoteAddressResolver implements ZLinkSpotRemoteAddressResolver {
    private final RedisRoomRouteStore routes;

    public RedisSpotRemoteAddressResolver(RedisRoomRouteStore routes) {
        this.routes = routes;
    }

    @Override
    public CompletionStage<ZLinkSpotRemoteAddress> resolveSpotRemoteAddressAsync(RoutingId spotRid) {
        RedisRoomRouteStore.RoomRoute route = routes.require(spotRid.toString());
        return CompletableFuture.completedFuture(new ZLinkSpotRemoteAddress(
            SampleNames.RouteChannel,
            RoutingId.from(route.ownerSpotNodeRid()),
            spotRid,
            ZLinkSpotKind.USER));
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/configuration/RedisSpotRemoteAddressResolver.kt
package systems.zlink.samples.kotlin.tictactoe.server.configuration

import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import systems.zlink.contracts.core.RoutingId
import systems.zlink.framework.spots.ZLinkSpotKind
import systems.zlink.framework.spots.ZLinkSpotRemoteAddress
import systems.zlink.framework.spots.ZLinkSpotRemoteAddressResolver

class RedisSpotRemoteAddressResolver(
    private val routeStore: RedisRoomRouteStore,
) : ZLinkSpotRemoteAddressResolver {
    override fun resolveSpotRemoteAddressAsync(spotRid: RoutingId): CompletionStage<ZLinkSpotRemoteAddress> {
        val route = routeStore.require(spotRid.toString())
        return CompletableFuture.completedFuture(
            ZLinkSpotRemoteAddress(
                SampleNames.RouteChannel,
                RoutingId.from(route.ownerSpotNodeRid),
                spotRid,
                ZLinkSpotKind.USER,
            ),
        )
    }
}
